File: tic_tac_php/board.php
<?php
interface IBoard {
    public function isValidPosition(int $row, int $col): bool;
    public function getCell(int $row, int $col): string;
    public function setCell(int $row, int $col, string $mark);
    public function display();
    public function isFull(): bool;
    public function hasHorizontalWin(): bool;
    public function hasVerticalWin(): bool;
    public function hasDiagonalWin(): bool;
    public function checkWinner(): bool;
    public function setRemainingCells();
}

class Board implements IBoard {
    const SIZE = 3;
    private $cells;

    public function __construct() {
        $this->cells = array_fill(0, self::SIZE, array_fill(0, self::SIZE, ''));
    }

    public function __get($name) {
        return $this->$name;
    }

    public function isValidPosition(int $row, int $col, bool $fullCheck = true): bool {
        if ($row < 0 || $col < 0 || $row >= self::SIZE || $col >= self::SIZE) {
            return false;
        }

        $cell = $this->cells[$row][$col];

        return $fullCheck ? empty($cell) : true;
    }

    public function getCell(int $row, int $col): string {
        if (!$this->isValidPosition($row, $col, false)) {
            throw new Exception("Invalid row or column value. Try again");
        }

        return $this->cells[$row][$col];
    }

    public function setCell(int $row, int $col, string $mark) {
        if (!$this->isValidPosition($row, $col)) {
            throw new Exception("Invalid row or column value. Try again");
        }

        $this->cells[$row][$col] = $mark;
    }

    public function display() {
        echo "\n     c1  c2  c3\n";
        echo "    ------------\n";
        
        foreach ($this->cells as $i => $row) {
            echo "r" . ($i + 1) . " |";
            foreach ($row as $j => $cell) {
                echo sprintf(" %1s |", $cell);
            }

            echo "\n   -------------\n";
        }

        print_nline(1);
    }

    public function isFull(): bool {
        foreach ($this->cells as $row) {
            if (in_array('', $row)) {
                return false;
            }
        }

        return true;
    }

    public function hasHorizontalWin(): bool {
        foreach ($this->cells as $row) {
            if (!in_array('', $row) && count(array_unique($row)) === 1) {
                return true;
            }
        }

        return false;
    }

    public function hasVerticalWin(): bool {
        foreach (array_map(null, ...$this->cells) as $col) {
            if (!in_array('', $col) && count(array_unique($col)) === 1) {
                return true;
            }
        }
        return false;
    }

    public function hasDiagonalWin(): bool {
        $mainDiagonal = [];
        $secDiagonal = [];

        for ($i = 0; $i < self::SIZE; $i++) {
            $mainDiagonal[] = $this->getCell($i, $i);
            $secDiagonal[] = $this->getCell($i, self::SIZE - $i - 1);
        }

        $main_match = !in_array('', $mainDiagonal) && count(array_unique($mainDiagonal)) === 1;
        $sec_match = !in_array('', $secDiagonal) && count(array_unique($secDiagonal)) === 1;

        return $main_match || $sec_match;
    }

    public function checkWinner(): bool {
        return $this->hasHorizontalWin() || $this->hasVerticalWin() || $this->hasDiagonalWin();
    }

    public function setRemainingCells() {
        foreach ($this->cells as $i => $row) {
            foreach ($row as $j => $cell) {
                if (empty($cell)) {
                    $this->setCell($i, $j, "-");
                }
            }
        }
    }


}

?>

File: tic_tac_rb/board.rb
class Board
    SIZE = 3
    attr_accessor :cells

    def initialize
        @cells = Array.new(SIZE) { Array.new(SIZE, '') }
    end

    def is_valid_position(row, col)
        if row < 0 or col < 0 or row >= SIZE or col >= SIZE
            return false
        end

        cell = @cells[row][col]

        return cell.strip.empty?
    end

    def get_cell(row, col)
        if not is_valid_position(row, col)
            raise IndexError, "Invalid row or column value. Try again"
        end
        return @cells[row][col]
    end

    def set_cell(row, col, mark)
        if not is_valid_position(row, col)
            raise IndexError, "Invalid row or column value. Try again"
        end
        @cells[row][col] = mark
    end

    def display
        puts "\n     c1  c2  c3"
        puts "    ------------"

        @cells.each_with_index do |row, i|
            print "r#{i+1} |"
            row.each_with_index do |cell, j|
                print sprintf(" %1s |", cell)
            end

            puts "\n   -------------"
        end

        puts
    end

    def is_full
        @cells.all? { |row| row.none?(&:empty?) }
    end

    def has_horizontal_win
        @cells.any? { |row| row.none?(&:empty?) and row.uniq.length == 1 }
    end

    def has_vertical_win
        @cells.transpose.any? { |col| col.none?(&:empty?) and col.uniq.length == 1 }
    end

    def has_diagonal_win
        main_diagonal = (0...SIZE).map { |i| @cells[i][i] }
        sec_diagonal = (0...SIZE).map { |i| @cells[i][SIZE - i - 1 ] }

        main_match = main_diagonal.none?(&:empty?) and main_diagonal.uniq.length == 1
        sec_match = sec_diagonal.none?(&:empty?) and sec_diagonal.uniq.length == 1

        main_match or sec_match
    end

    def check_winner
        has_horizontal_win or has_vertical_win or has_diagonal_win
    end

    def set_remaining_cells
        @cells.each_with_index do |row, i|
            row.each_with_index do |cell, j|
                if cell.empty?
                    set_cell(i, j, "-")
                end
            end
        end
    end

end